### PR TITLE
feat(similarity-embedding): Create backfill script for inserting records

### PR DESCRIPTION
**Pull request by jangjodi**
Create script to backfill seer grouping records
Create function to call seer bulk insert endpoint (https://github.com/getsentry/seer/pull/480) in batches of 20 groups with a 10 second timeout
Add seer bulk insert function
Modify get_grouping_info function to take in an optional event

The script will be triggered using an endpoint, which will be in a follow-up PR. This is because the script will only be run for the s4s project for now, thus we only need to run it once.

closes https://github.com/getsentry/getsentry/issues/13420

### CODE REVIEW

File: src/sentry/seer/utils.py [hunk 1]
```diff
@@ -40,14 +46,7 @@ class BreakpointResponse(TypedDict):
...
 
-seer_staging_connection_pool = connection_from_url(
-    settings.SEER_AUTOFIX_URL,
-    retries=Retry(
-        total=5,
-        status_forcelist=[408, 429, 502, 503, 504],
-    ),
-    timeout=settings.ANOMALY_DETECTION_TIMEOUT,
-)
+seer_staging_connection_pool = connection_from_url(settings.SEER_AUTOFIX_URL)
```
Comment (jangjodi) [inline src/sentry/seer/utils.py:49]: Remove the automatic retries to seer for similarity endpoints

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+                    if group.data.get("metadata"):
+                        group.data["metadata"].update({"has_embeddings_record_v1": True})
+                    else:
+                        group.data["metadata"] = {"has_embeddings_record_v1": True}
+                num_groups_records_created += Group.objects.bulk_update(groups, ["data"])
+
+    return num_groups_records_created
+
+
+def lookup_group_data_stacktrace_bulk_with_fallback(
```
Comment (jangjodi) [inline src/sentry/tasks/backfill_seer_grouping_records.py:137]: this function and the following functions were based off of this [backfill script](https://github.com/getsentry/getsentry/blob/b2689f4e98cb5d641ecdda4605e9cf84932b2a67/bin/backfill_perf_issue_events_issue_platform.py#L476)

File: src/sentry/seer/utils.py (hunk 1 above)
Comment (snigdhas) [reply src/sentry/seer/utils.py:49]: do we still want the timeout here?

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+
+    project_id, organization_id = project.id, project.organization.id
+    time = datetime.now()
+    events_entity = Entity("events", alias="events")
+    redis_client = redis.redis_clusters.get(settings.SENTRY_MONITORS_REDIS_CLUSTER)
+    last_processed_id = int(redis_client.get(LAST_PROCESSED_REDIS_KEY) or 0)
+    for group_id_batch in chunked(
+        RangeQuerySetWrapper(
+            Group.objects.filter(
+                project_id=project_id, id__gt=last_processed_id, type=1
```
Comment (jangjodi) [inline src/sentry/tasks/backfill_seer_grouping_records.py:74]: make type the GroupType.id

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+            result_value_getter=lambda item: item,
+        ),
+        ITERATOR_CHUNK,
+    ):
+        query = Query(
+            match=events_entity,
+            select=[
+                Column("group_id"),
+                Function("max", [Column("event_id")], "event_id"),
+                Column("message"),
```
Comment (jangjodi) [inline src/sentry/tasks/backfill_seer_grouping_records.py:85]: double check this is 1-1

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+            orderby=[OrderBy(Column("group_id"), Direction.ASC)],
+        )
+
+        request = Request(
+            dataset=Dataset.Events.value,
+            app_id=Referrer.GROUPING_RECORDS_BACKFILL_REFERRER.value,
+            query=query,
+            tenant_ids={
+                "referrer": Referrer.GROUPING_RECORDS_BACKFILL_REFERRER.value,
+                "organization_id": organization_id,
```
Comment (jangjodi) [inline src/sentry/tasks/backfill_seer_grouping_records.py:105]: shouldn't trigger ratelimit for orgs - see if it can be done just for referrer

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+                        and not (
+                            event.group.data.get("metadata")
+                            and event.group.data["metadata"].get("has_embeddings_record_v1")
+                        )
+                    ):
+                        grouping_info = get_grouping_info(
+                            None, project=project, event_id=event.event_id, event=event
+                        )
+                        stacktrace_string = get_stacktrace_string(grouping_info)
+                        if stacktrace_string == "":
```
Comment (jangjodi) [inline src/sentry/tasks/backfill_seer_grouping_records.py:228]: if groups are invalid (no stacktrace) we should not try again later

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+                    if (
+                        event.data
+                        and event.data.get("exception")
+                        and event.group
+                        and not (
+                            event.group.data.get("metadata")
+                            and event.group.data["metadata"].get("has_embeddings_record_v1")
+                        )
+                    ):
+                        grouping_info = get_grouping_info(
```
Comment (jangjodi) [inline src/sentry/tasks/backfill_seer_grouping_records.py:224]: follow up with any ramifications if grouping config changes

**Review by wedamija — COMMENTED** [review]
Is the idea that this only needs to be run on prod? If so, it might make sense to move to getsentry

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+    stacktrace_list: list[str]
+
+
+@instrumented_task(
+    name="sentry.tasks.backfill_seer_grouping_records",
+    queue="default",
+    max_retries=0,
+    silo_mode=SiloMode.REGION,
+)
+def backfill_seer_grouping_records(project: Project) -> int:
```
[the remark this answers is not in the document]
Comment (wedamija) [reply src/sentry/tasks/backfill_seer_grouping_records.py:61]: Nm read the description

Comment (jangjodi) [inline src/sentry/tasks/backfill_seer_grouping_records.py:61]: take in args and kwargs for backwards compatibility

**Comment by jangjodi** [comment]
> Is the idea that this only needs to be run on prod? If so, it might make sense to move to getsentry

Thanks, I've opened a separate getsentry PR [here](https://github.com/getsentry/getsentry/pull/13606)

**Comment by jangjodi** [comment]
> Is the idea that this only needs to be run on prod? If so, it might make sense to move to getsentry

we're going to be using this script or a version of it in post process eventually, so I'm going to keep it in sentry

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+                group_hashes = GroupHash.objects.filter(
+                    project_id=project_id, id__in=group_id_batch
+                )
+                group_hashes_dict = {
+                    group_hash.group_id: group_hash.hash for group_hash in group_hashes
+                }
+                data = lookup_group_data_stacktrace_bulk_with_fallback(
+                    project, rows, group_id_message_batch, group_hashes_dict
+                )
+                response = post_bulk_grouping_records(
```
Comment (JoshFerge) [inline src/sentry/tasks/backfill_seer_grouping_records.py:130]: can we get timing for this function? can add the metrics.wraps decorator to it. like https://github.com/getsentry/sentry/blob/78b8cbeb9380058577e90c5d85f7c211ed2b1d21/src/sentry/replays/usecases/ingest/__init__.py#L178

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+            invalid_event_ids,
+            GroupStacktraceData(data=group_data, stacktrace_list=stacktrace_strings),
+        )
+
+
+# add timing call for total time of task, nodestore, snuba, seer
+def lookup_group_data_stacktrace_single(
+    project: Project, event_id: str, group_id: int, message: str, hash: str
+) -> tuple[CreateGroupingRecordData | None, str]:
+    with metrics.timer(f"{BACKFILL_NAME}.lookup_event_single", sample_rate=1.0):
```
Comment (JoshFerge) [inline src/sentry/tasks/backfill_seer_grouping_records.py:282]: can use @metrics.wraps on the function here instead of the timer context to avoid the indentation

**Review by JoshFerge — APPROVED** [review]
looks good. a couple of metric additions would be good I think.

File: src/sentry/api/endpoints/event_grouping_info.py [hunk 11]
```diff
@@ -8,17 +9,20 @@
...
 
 logger = logging.getLogger(__name__)
 
 
-def get_grouping_info(config_name: str | None, project: Project, event_id: str):
-    event = eventstore.backend.get_event_by_id(project.id, event_id)
+def get_grouping_info(
+    config_name: str | None, project: Project, event_id: str, event: Event | None = None
+):
+    event = event if event else eventstore.backend.get_event_by_id(project.id, event_id)
```
Comment (lobsterkatie) [inline src/sentry/api/endpoints/event_grouping_info.py:25]: To make rebasing easier, I pulled this out into a separate PR: https://github.com/getsentry/sentry/pull/69932.

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+from typing import Any, TypedDict
+
+import sentry_sdk
+from django.conf import settings
+from google.api_core.exceptions import DeadlineExceeded, ServiceUnavailable
+from snuba_sdk import Column, Condition, Entity, Function, Op, Query, Request
+from snuba_sdk.orderby import Direction, OrderBy
+
+from sentry import features, nodestore
+from sentry.api.endpoints.event_grouping_info import get_grouping_info
```
Comment (lobsterkatie) [inline src/sentry/tasks/backfill_seer_grouping_records.py:13]: As of https://github.com/getsentry/sentry/pull/68579, `get_grouping_info` now lives in the `grouping` module.

```suggestion
from sentry.grouping.grouping_info import get_grouping_info
```

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+            if node_id in node_id_to_group_data:
+                event_id, group_id = (
+                    node_id_to_group_data[node_id][0],
+                    node_id_to_group_data[node_id][1],
+                )
+                event = Event(event_id=event_id, project_id=project_id, group_id=group_id)
+                event.data = data
+                if event and event.data and event.data.get("exception") and hashes.get(group_id):
+                    grouping_info = get_grouping_info(
+                        None, project=project, event_id=event.event_id, event=event
```
Comment (lobsterkatie) [inline src/sentry/tasks/backfill_seer_grouping_records.py:251]: As of https://github.com/getsentry/sentry/pull/69932, we only have to pass in the event.

```suggestion
                        None, project=project, event=event
```

File: src/sentry/features/temporary.py [hunk 14]
```diff
@@ -275,6 +275,7 @@ def register_temporary_features(manager: FeatureManager):
     manager.add("projects:similarity-embeddings", ProjectFeature, FeatureHandlerStrategy.INTERNAL)
     manager.add("projects:similarity-embeddings-grouping", ProjectFeature, FeatureHandlerStrategy.INTERNAL)
     manager.add("projects:similarity-embeddings-metadata", ProjectFeature, FeatureHandlerStrategy.INTERNAL)
+    manager.add("projects:similarity-embeddings-remove-seer-grouping-record-table-init", ProjectFeature, FeatureHandlerStrategy.INTERNAL)
```
Comment (jangjodi) [inline src/sentry/features/temporary.py:278]: adding a feature flag to allow s4s to drop the existing (incorrect) records table and not re-initialize it so that we can try backfilling it with the correct information [here](https://github.com/getsentry/seer/blob/506aae671c35396b8c928073a3ad3a1d55d6ae79/src/seer/grouping/grouping.py#L98). we will eventually want to remove this janky initialization logic for all orgs, but we want to start with s4s

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+                        }
+                    else:
+                        group.data["metadata"] = {
+                            "embeddings_info": {
+                                "nn_model_version": 1,
+                                "group_hash": json.dumps([group_hashes_dict[group.id]]),
+                            }
+                        }
+                num_groups_records_created += Group.objects.bulk_update(groups, ["data"])
+            time.sleep(ITERATOR_CHUNK * SEER_BACKFILL_DELAY_PER_RECORD)
```
Comment (wedamija) [inline src/sentry/tasks/backfill_seer_grouping_records.py:162]: How long are you expecting this task to take? If it's pretty long you might hit a timeout.

One option here is to refire the task and pass the id you just completed and continue on

File: src/sentry/features/temporary.py (hunk 14 above)
Comment (JoshFerge) [reply src/sentry/features/temporary.py:278]: talked offline, it is best to do database updates like this through migrations or a script instead of in realtime in an endpoint if possible

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 16]
```diff
@@ -0,0 +1,335 @@
...
+from sentry.snuba.referrer import Referrer
+from sentry.tasks.base import instrumented_task
+from sentry.utils import json, metrics, redis
+from sentry.utils.iterators import chunked
+from sentry.utils.query import RangeQuerySetWrapper
+from sentry.utils.safe import get_path
+from sentry.utils.snuba import bulk_snuba_queries
+
+ITERATOR_CHUNK = 20
+SEER_BACKFILL_DELAY_PER_RECORD = 0.1
```
Comment (JoshFerge) [inline src/sentry/tasks/backfill_seer_grouping_records.py:37]: let's consider making these options if in the future we want to quickly change these. but fine for now.

**Review by JoshFerge — APPROVED** [review]

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 17]
```diff
@@ -0,0 +1,342 @@
...
+
+
+@instrumented_task(
+    name="sentry.tasks.backfill_seer_grouping_records",
+    queue="default",
+    max_retries=0,
+    silo_mode=SiloMode.REGION,
+)
+@metrics.wraps(f"{BACKFILL_NAME}.task")
+def backfill_seer_grouping_records(project: Project, *args: Any, **kwargs: Any) -> None:
```
Comment (wedamija) [inline src/sentry/tasks/backfill_seer_grouping_records.py:62]: Will we eventually run this for all projects?

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 18]
```diff
@@ -0,0 +1,342 @@
...
+                project_id=project.id, id__gt=last_processed_id, type=ErrorGroupType.type_id
+            ).values_list("id", "message", "data"),
+            result_value_getter=lambda item: item[0],
+        ),
+        ITERATOR_CHUNK,
+    ):
+        backfill_chunked_seer_grouping_records.apply_async(
+            args=[group_id_message_data_batch, project.id],
+            countdown=ITERATOR_CHUNK * SEER_BACKFILL_DELAY_PER_RECORD,
+        )
```
Comment (wedamija) [inline src/sentry/tasks/backfill_seer_grouping_records.py:81]: This way will work ok, although one concern I have is that with a lot of schedule tasks pending we might take up a lot of memory in rabbit/celery.  Here's another option

```
def backfill_seer_grouping_records(project: Project, last_processed_id=None, *args: Any, **kwargs: Any) -> None:
    ...
    if last_processed_id is None:
        last_processed_id = int(redis_client.get(LAST_PROCESSED_REDIS_KEY) or 0)

    MAX_CHUNKS = X
    for i, chunk in enumerate(chunked):
        # Call seer function
        last_processed_id = group_id_message_data_batch[-1][0]
        redis_client.set(f"{LAST_PROCESSED_REDIS_KEY}", last_processed_id, ex=60 * 60 * 24 * 7)
        if i >= MAX_CHUNKS:
            backfill_seer_grouping_records.delay(project_id, last_processed_id=last_processed_id)
            return
```

Since we don't actually want to run these in parallel, it might be better to handle it this way.

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 19]
```diff
@@ -0,0 +1,342 @@
...
+class GroupStacktraceData(TypedDict):
+    data: list[CreateGroupingRecordData]
+    stacktrace_list: list[str]
+
+
+@instrumented_task(
+    name="sentry.tasks.backfill_seer_grouping_records",
+    queue="default",
+    max_retries=0,
+    silo_mode=SiloMode.REGION,
```
Comment (JoshFerge) [inline src/sentry/tasks/backfill_seer_grouping_records.py:59]: let's set the `soft_time_limit` to 15 minutes and `time_limit` to 15 minutes and 5 seconds

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 20]
```diff
@@ -0,0 +1,342 @@
...
+
+
+class GroupStacktraceData(TypedDict):
+    data: list[CreateGroupingRecordData]
+    stacktrace_list: list[str]
+
+
+@instrumented_task(
+    name="sentry.tasks.backfill_seer_grouping_records",
+    queue="default",
```
Comment (JoshFerge) [inline src/sentry/tasks/backfill_seer_grouping_records.py:57]: immediately after we get this merged, lets put this on its own celery queue.

File: src/sentry/tasks/backfill_seer_grouping_records.py (hunk 17 above)
Comment (JoshFerge) [inline src/sentry/tasks/backfill_seer_grouping_records.py:62]: irl comment: have this be a project_id in instead of a `Project` as w/ celery we need to use primitives

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 21]
```diff
@@ -0,0 +1,338 @@
...
+        .order_by("id")[:BATCH_SIZE]
+    )
+    group_id_message_batch = {
+        group_id: message
+        for (group_id, message, data) in group_id_message_data_batch
+        if not get_path(data, "metadata", "embeddings_info", "nn_model_version")
+    }
+    batch_len = len(group_id_message_batch)
+    if batch_len == 0:
+        return
```
Comment (wedamija) [inline src/sentry/tasks/backfill_seer_grouping_records.py:87]: This could cause this task to end early if all groups in the batch have `nn_model_version` set. Do we actually want to halt here? What if later groups don't have it set, or if we're wanting to re-run this?

This probably makes sense to do as part of `group_id_message_data_batch`?

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 22]
```diff
@@ -0,0 +1,338 @@
...
+
+    group_id_batch = list(group_id_message_batch.keys())
+    time_now = datetime.now()
+    events_entity = Entity("events", alias="events")
```
Comment (JoshFerge) [inline src/sentry/tasks/backfill_seer_grouping_records.py:91]: follow up: should this be the errors entity?

File: src/sentry/tasks/backfill_seer_grouping_records.py [hunk 23]
```diff
@@ -0,0 +1,338 @@
...
+def backfill_seer_grouping_records(
+    project_id: int, last_processed_id: int, *args: Any, **kwargs: Any
+) -> None:
+    project = Project.objects.get_from_cache(id=project_id)
+    if not features.has("projects:similarity-embeddings-grouping", project):
+        return
+
+    redis_client = redis.redis_clusters.get(settings.SENTRY_MONITORS_REDIS_CLUSTER)
+    if last_processed_id is None:
+        last_processed_id = int(redis_client.get(LAST_PROCESSED_REDIS_KEY) or 0)
```
Comment (JoshFerge) [inline src/sentry/tasks/backfill_seer_grouping_records.py:71]: follow up: use project_id in this redis key

Comment (JoshFerge) [reply src/sentry/tasks/backfill_seer_grouping_records.py:71]: follow up: should we set a TTL on this redis key?

**Review by JoshFerge — APPROVED** [review]

File: src/sentry/features/temporary.py [hunk 24]
```diff
@@ -276,6 +276,7 @@ def register_temporary_features(manager: FeatureManager):
     manager.add("projects:similarity-embeddings", ProjectFeature, FeatureHandlerStrategy.INTERNAL)
     manager.add("projects:similarity-embeddings-grouping", ProjectFeature, FeatureHandlerStrategy.INTERNAL)
     manager.add("projects:similarity-embeddings-metadata", ProjectFeature, FeatureHandlerStrategy.INTERNAL)
+    manager.add("projects:similarity-embeddings-backfill", ProjectFeature, FeatureHandlerStrategy.INTERNAL)
```
[the remark this answers is not in the document]
Comment (jangjodi) [reply src/sentry/features/temporary.py]: ah yes good catch, I changed it to options

**Review by JoshFerge — APPROVED** [review]
